File: rest-service/src/main/java/org/veupathdb/service/multiblast/model/blast/TemplateType.java
package org.veupathdb.service.multiblast.model.blast;

import java.util.Optional;

public enum TemplateType
{
  Coding("coding"),
  Optimal("optimal"),
  Both("coding_and_optimal"),
  ;

  private final String value;

  TemplateType(String value) {
    this.value = value;
  }

  public String getValue() {
    return value;
  }

  public static Optional<TemplateType> fromString(String name) {
    for (var e : values())
      if (e.value.equals(name))
        return Optional.of(e);

    return Optional.empty();
  }

  public static TemplateType unsafeFromString(String name) {
    return fromString(name).orElseThrow(IllegalArgumentException::new);
  }
}

File: common/blast-types/src/main/kotlin/org/veupathdb/lib/blast/blastn/fields/TemplateType.kt
package org.veupathdb.lib.blast.blastn.fields

import com.fasterxml.jackson.databind.node.ObjectNode
import org.veupathdb.lib.blast.common.FlagDust
import org.veupathdb.lib.blast.common.FlagTemplateType
import org.veupathdb.lib.blast.serial.BlastField
import org.veupathdb.lib.blast.util.*


internal fun ParseTemplateType(js: ObjectNode) =
  js.optString(FlagTemplateType) { TemplateType(parseEnum(it)) }
    ?: TemplateType()


/**
 * -template_type `<String>`
 *
 * Permissible Values:
 * * coding
 * * coding_and_optimal
 * * optimal
 *
 * Discontiguous MegaBLAST template type
 */
@JvmInline
value class TemplateType(val value: TemplateTypeType = TemplateTypeType.None)
  : BlastField
{
  override val isDefault get() = value == TemplateTypeType.None

  override val name: String
    get() = FlagTemplateType

  override fun appendJson(js: ObjectNode) =
    js.put(isDefault, FlagTemplateType, value.value)

  override fun appendCliSegment(cli: StringBuilder) =
    cli.append(isDefault, FlagTemplateType, value.value)

  override fun appendCliParts(cli: MutableList<String>) =
    cli.add(isDefault, FlagTemplateType, value.value)
}


private fun parseEnum(value: String): TemplateTypeType {
  return when(value) {
    "coding"             -> TemplateTypeType.Coding
    "coding_and_optimal" -> TemplateTypeType.CodingAndOptimal
    "optimal"            -> TemplateTypeType.Optimal
    else                 -> throw IllegalArgumentException("Invalid $FlagTemplateType value: $value")
  }
}


enum class TemplateTypeType {
  Coding,
  CodingAndOptimal,
  Optimal,
  None;

  val value get() = when(this) {
    Coding           -> "coding"
    CodingAndOptimal -> "coding_and_optimal"
    Optimal          -> "optimal"
    None             -> "none"
  }
}
